tests/helics/common/SmallBufferTests.cpp

```diff
@@ -0,0 +1,478 @@
+    EXPECT_EQ(sb1[13], std::byte{'a'});
+    SmallBuffer sb2(std::string(36223, 'b'));
+    EXPECT_EQ(sb2.size(), 36223);
+    sb2 = std::move(sb1);
+    EXPECT_EQ(sb2[11], std::byte{'a'});
+}
+
+TEST(small_buffer_tests, move_assign_self)
+{
+    SmallBuffer sb1(std::string(36214, 'e'));
+    EXPECT_EQ(sb1.size(), 36214);
+    sb1 = std::move(sb1);
+    EXPECT_EQ(sb1[11], std::byte{'e'});  // NOLINT
+}
+
+TEST(small_buffer_tests, buffer_transfer)
+{
+    SmallBuffer sb1;
+    auto* buffer = new std::byte[5000];
+
+    sb1.moveAssign(buffer, 4567, 5000);
+    EXPECT_EQ(sb1.size(), 4567U);
+    EXPECT_EQ(sb1.capacity(), 5000);
+    sb1[27] = std::byte{15};
+
+    EXPECT_EQ(buffer[27], std::byte{15});
+    // the SMallbuffer should take care of deletion
+}
+
+TEST(small_buffer_tests, buffer_transfer_full)
+{
+    SmallBuffer sb1(std::string(2354, 'b'));
+    auto* buffer = new std::byte[5000];
+
+    sb1.moveAssign(buffer, 4567, 5000);
+    EXPECT_EQ(sb1.size(), 4567U);
+    EXPECT_EQ(sb1.capacity(), 5000);
+    sb1[27] = std::byte{15};
+
+    EXPECT_EQ(buffer[27], std::byte{15});
+    // the SMallbuffer should take care of deletion
+}
+
+TEST(small_buffer_tests, buffer_transfer_self_assign)
+{
+    SmallBuffer sb1(std::string(2354, 'b'));
+
+    sb1.moveAssign(sb1.data(), 2314, 2354);
+    EXPECT_EQ(sb1.size(), 2314U);
+    EXPECT_GE(sb1.capacity(), 2354);
+    sb1[27] = std::byte{15};
+
+    EXPECT_EQ(sb1[27], std::byte{15});
+    EXPECT_EQ(sb1[20], std::byte{'b'});
+    // the SMallbuffer should take care of deletion
+}
+
+TEST(small_buffer_tests, buffer_borrow)
+{
+    SmallBuffer sb1;
+    std::vector<std::byte> buffer;
+    buffer.resize(5000);
+
+    sb1.spanAssign(buffer.data(), 4567, 5000);
+    EXPECT_EQ(sb1.size(), 4567U);
+    EXPECT_EQ(sb1.capacity(), 5000);
+    sb1[27] = std::byte{15};
+
+    EXPECT_EQ(buffer[27], std::byte{15});
+    // the Smallbuffer should not delete the object
+}
+
+TEST(small_buffer_tests, buffer_borrow_full)
+{
+    SmallBuffer sb1(std::string(2354, 'b'));
+    std::vector<std::byte> buffer;
+    buffer.resize(5000);
+
+    sb1.spanAssign(buffer.data(), 4567, 5000);
+    EXPECT_EQ(sb1.size(), 4567U);
+    EXPECT_EQ(sb1.capacity(), 5000);
+    sb1[27] = std::byte{15};
+
+    EXPECT_EQ(buffer[27], std::byte{15});
+    // the Smallbuffer should not delete the object
+}
+
+TEST(small_buffer_tests, buffer_borrow_self_assign)
+{
+    SmallBuffer sb1(std::string(2354, 'b'));
+
+    sb1.spanAssign(sb1.data(), 1986, 2000);
+    EXPECT_EQ(sb1.size(), 1986U);
+    EXPECT_GE(sb1.capacity(), 2000);
+    sb1[27] = std::byte{15};
+
+    EXPECT_EQ(sb1[27], std::byte{15});
+    EXPECT_EQ(sb1[20], std::byte{'b'});
+    // this should trigger self assignment detection
+}
+
+TEST(small_buffer_tests, assign)
+{
+    SmallBuffer sb1;
+    std::string t1(3634, 'g');
+    sb1.assign(t1.data(), t1.data() + t1.size());
+    EXPECT_EQ(t1.size(), sb1.size());
+    EXPECT_EQ(sb1[3333], std::byte{'g'});
+}
```

Comment:
What happens with an assignment of size zero? 
```
    sb1.assign(t1.data(), t1.data());
    EXPECT_EQ(sb1.size(), 0);
```
Would be a valid test here, but I do not believe it would be a meaningful buffer, and I believe attempts to access the buffer object would result in a segfault.

Reply:
as far as I can this test works fine.  If done initially the contained buffer is active,  otherwise it doesn't change the existing buffer.  So unless someone passed in a void buffer this will always result in a valid buffer and not produce a seg fault.

Reply:
You are correct, I was attempting to access the buffer index in an invalid way, and caused the segfault myself.